homeassistant/components/xiaomi_miio/number.py

```diff
@@ -0,0 +1,214 @@
+        device_features = 0
+        entity_class = None
+
+        _LOGGER.debug("Initializing with host %s (token %s...)", host, token[:5])
+        if model in [MODEL_AIRHUMIDIFIER_CA1, MODEL_AIRHUMIDIFIER_CB1]:
+            device_features = FEATURE_FLAGS_AIRHUMIDIFIER_CA_AND_CB
+            entity_class = XiaomiAirHumidifierNumber
+        elif model in [MODEL_AIRHUMIDIFIER_CA4]:
+            device_features = FEATURE_FLAGS_AIRHUMIDIFIER_CA4
+            entity_class = XiaomiAirHumidifierNumber
+        elif model in MODELS_HUMIDIFIER:
+            device_features = FEATURE_FLAGS_AIRHUMIDIFIER
+            entity_class = XiaomiAirHumidifierNumber
+        else:
+            return
+
+        for feature in NUMBER_TYPES:
+            number = NUMBER_TYPES[feature]
+            if feature & device_features and feature in NUMBER_TYPES:
+                entities.append(
+                    entity_class(
+                        f"{config_entry.title} {number.name}",
+                        device,
+                        config_entry,
+                        f"{number.short_name}_{config_entry.unique_id}",
+                        number,
+                        coordinator,
+                    )
+                )
+
+    async_add_entities(entities)
+
+
+class XiaomiNumber(XiaomiCoordinatedMiioEntity, NumberEntity):
+    """Representation of a generic Xiaomi attribute selector."""
+
+    def __init__(self, name, device, entry, unique_id, number, coordinator):
+        """Initialize the generic Xiaomi attribute selector."""
+        super().__init__(name, device, entry, unique_id, coordinator)
+        self._state = None
+        self._attr_icon = number.icon
+        self._attr_unit_of_measurement = number.unit_of_measurement
+        self._attr_min_value = number.min
+        self._attr_max_value = number.max
+        self._attr_step = number.step
+        self._supported_features = 0
+        self._device_features = 0
+        self._state_attrs = {}
+        self._controller = number
+        self._enum_class = None
+        self._attributes = None
+        self._value = None
+
+    @property
+    def value(self):
+        """Return the current numbered value."""
+        return self._value
+
+    @property
+    def available(self):
+        """Return the number controller availability."""
+        if (
+            super().available
+            and not self.coordinator.data.is_on
+            and not self._controller.available_with_device_off
+        ):
+            return False
+        return super().available
+
+    @staticmethod
+    def _extract_value_from_attribute(state, attribute):
+        value = getattr(state, attribute)
+        if isinstance(value, Enum):
+            return value.value
+
+        return value
+
+    async def async_set_value(self, value):
+        """Set an option of the miio device."""
+        if not self.available:
+            return
+        if (
+            self.min_value
```

Comment:
How, If I pass an invalid value. The GUI will show red (not blue) but it still will pass the value if I submit it.

Reply:
The service handler should raise a `ValueError`.

Reply:
So, if this does not happen this is a possible bug in the `number` integration? Because no `ValueError` is raised. I am seeing warning the the logs I pass an invalid value.
```
2021-06-28 09:55:44 WARNING (MainThread) [homeassistant.components.xiaomi_miio.number] Value 190.0 not a valid zhimi.humidifier.ca4 Motor Speed within the range 200 - 2000
```

Reply:
I have updated the code to raise a ValueError now instead of logging a warning

Reply:
My suggestion was that we in a separate PR add this check to the number integration. There is currently no check.

Reply:
> 
> 
> My suggestion was that we in a separate PR add this check to the number integration. There is currently no check.

I cannot completely see hoe we should do that. We have a similar issue with the led_brightness PR. I also added a check there.
What whould be the scope of this check PR?

Reply:
Add an intermediate service handler in the number integration that takes the entity and service call as parameters. Apply the min/max check and raise a `ValueError` if not ok, otherwise call the entity `async_set_value` method.

See the last example here for an intermediate service handler:
https://developers.home-assistant.io/docs/dev_101_services#entity-services

Reply:
And yes, this is a separate task, not related to this PR here for now.